spec/ics-3-connection-semantics/README.md

```diff
@@ -0,0 +1,245 @@
+
+The Merkle proof `M_kvh` is a well-defined concept in the blockchain space, and provides a compact proof that the key value pair `(k, v)` is consistent with a Merkle root stored in `H_h`. Handling the case where `k` is not in the store requires a separate proof of non-existence, which is not supported by all Merkle stores. Thus, we define the proof only as a proof of existence. There is no valid proof for missing keys, and we design the algorithm to work without it. 
+
+Blockchains supporting IBC must implement Merkle proof verification:
+
+`valid(H_h, M_kvh) ⇒ true | false`
+
+#### Lifecycle
+
+##### Opening a connection
+
+All proofs require an initial `H_h` and `C_h` for some `h`, where `dt(now, H_h) < P`.
+
+Establishing a bidirectional initial root-of-trust between the two blockchains (`A` to `B` and `B` to `A`) — `H_ah` and `C_ah` stored on chain `B`, and `H_bh` and `C_bh` stored on chain `A` — is necessary before any IBC packets can be sent.
+
+Any header may be from a malicious chain (e.g. shadowing a real chain state with a fake validator set), so a subjective decision is required before establishing a connection. This can be performed permissionlessly, in which case users later utilizing the IBC channel must check the root-of-trust themselves, or authorized by on-chain governance for additional assurance.
+
+*THREE WAY HANDSHAKE OUTLINE*
+
+###### OPENTRY
+
+First datagram, A -> B
+
+Provided input:
+- Commitment to root-of-trust for B
+
+On A:
+- Reserve connection slot
+- Prove reserved connection slot (is this necessary?)
+- Prove commitment to root-of-trust for B
+
+Data:
+- Root-of-trust for A
+- Connection slot identifier on A
+- Desired connection slot identifier on B
+- Proof of reserved connection slot identifier on A
+
+On B:
+- Check slot identifier
+- Ensure no existing root-of-trust for A
+- Store root-of-trust for A ~= connection identifier
+
+Timeout (committing on B) = stored RoT deleted, connection slot unreserved.
+
+Forgery: Any A' could commit such a packet on B. Only guarantee: non-duplication.
+
+###### OPENACK
+
+Second datagram, B -> A
+
+On B:
+- Prove stored root-of-trust for A in desired connection slot identifier
+
+Data:
+- Root-of-trust for B
+- Proof of stored root-of-trust for A in desired connection slot identifier
+
+On A:
+- Store root-of-trust for B ~= connection identifier
+- Connection now "open" on A
+
+Timeout (committing on A)
+- On B: Slot unreserved, root-of-trust for A deleted
+- On A: Connection timeout frees slot *OR* second datagram to prove unreserved slot, termined handshake can then free slot on A (eventually)
+
+Forgery: Any B' could commit ACK packet on A (unless RoT-hash is included; alternative version). Only guarantee: non-duplication.
```

Comment:
How? Didn't we already commit to B's root-of-trust?

Reply:
The handshake protocol has been improved and this is no longer possible.